Add more smart pointers to RadioInputType.cpp

abigailfox opened the pull request:
#### 8642253dec6d74b77c7072b9dfae5e1e8feb7407
<pre>
Add more smart pointers to RadioInputType.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=272118">https://bugs.webkit.org/show_bug.cgi?id=272118</a>
<a href="https://rdar.apple.com/125877697">rdar://125877697</a>

Reviewed by Ryosuke Niwa.

Add more smart pointers to RadioInputType.cpp

* Source/WebCore/html/RadioInputType.cpp:
(WebCore::RadioInputType::valueMissing const):
(WebCore::RadioInputType::willUpdateCheckedness):
(WebCore::RadioInputType::handleKeydownEvent):
(WebCore::RadioInputType::handleKeyupEvent):
(WebCore::RadioInputType::isKeyboardFocusable const):
(WebCore::RadioInputType::shouldSendChangeEventAfterCheckedChanged):
(WebCore::RadioInputType::willDispatchClick):
(WebCore::RadioInputType::didDispatchClick):
(WebCore::RadioInputType::matchesIndeterminatePseudoClass const):

Canonical link: <a href="https://commits.webkit.org/277319@main">https://commits.webkit.org/277319@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/865f2899157aa477084072fd32214c40019550aa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47059 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26229 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49697 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49742 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43106 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49366 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31434 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Checked out pull request; Running compile-webkit") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23695 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38329 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47640 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/31434 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/14/builds/49697 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19636 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/31434 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/14/builds/49697 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5104 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/31434 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/14/builds/49697 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51618 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22081 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18455 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45622 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23361 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/14/builds/49697 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44632 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10425 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24143 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23079 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/html/RadioInputType.cpp:
```diff
@@ -46,20 +46,20 @@ const AtomString& RadioInputType::formControlType() const
-    auto& name = element()->name();
-    if (auto* buttons = element()->radioButtonGroups())
-        return !buttons->checkedButtonForGroup(name) && buttons->isInRequiredGroup(*element());
+    ASSERT(protectedElement());
```
Instead of keep calling protectedElement(), can we store it in a local variable?
e.g.
```
RefPtr element = this->element();
```

rniwa commented on Source/WebCore/html/RadioInputType.cpp:
```diff
@@ -124,7 +124,8 @@ auto RadioInputType::handleKeydownEvent(KeyboardEvent& event) -> ShouldCallBaseE
         return ShouldCallBaseEventHandler::Yes;
 
-    ASSERT(element());
+    RefPtr elementRef = element();
```
I think we should call this variable just "element" and replace `element()` with `this->element()`

rniwa commented on Source/WebCore/html/RadioInputType.cpp:
```diff
@@ -166,10 +167,10 @@ void RadioInputType::handleKeyupEvent(KeyboardEvent& event)
     // If an unselected radio is tabbed into (because the entire group has nothing
     // checked, or because of some explicit .focus() call), then allow space to check it.
-    if (element()->checked()) {
+    if (protectedElement()->checked()) {
```
Can we store element in a local variable instead?

rniwa commented on Source/WebCore/html/RadioInputType.cpp:
```diff
@@ -215,10 +216,10 @@ void RadioInputType::willDispatchClick(InputElementClickState& state)
 
-    state.checked = element()->checked();
-    state.checkedRadioButton = element()->checkedRadioButtonForGroup();
+    state.checked = protectedElement()->checked();
```
Ditto about storing it in a local variable instead of keep calling protectedElement.

rniwa commented on Source/WebCore/html/RadioInputType.cpp:
```diff
@@ -228,11 +229,11 @@ void RadioInputType::didDispatchClick(Event& event, const InputElementClickState
         auto& button = state.checkedRadioButton;
         ASSERT(element());
-        if (button && button->isRadioButton() && button->form() == element()->form() && button->name() == element()->name())
+        if (button && button->isRadioButton() && button->form() == protectedElement()->form() && button->name() == protectedElement()->name())
```
Ditto about strong element in a local variable.

cdumez commented on Source/WebCore/html/RadioInputType.cpp:
```diff
@@ -96,11 +97,11 @@ void RadioInputType::willUpdateCheckedness(bool nowChecked, WasSetByJavaScript)
 {
     if (!nowChecked)
         return;
-    if (element()->radioButtonGroups()) {
```
Seems we should have a local variable for the element to avoid calling `protectedElement()` twice and causing unnecessary refcounting churn.

cdumez commented on Source/WebCore/html/RadioInputType.cpp:
```diff
@@ -124,19 +125,20 @@ auto RadioInputType::handleKeydownEvent(KeyboardEvent& event) -> ShouldCallBaseE
     // We can only stay within the form's children if the form hasn't been demoted to a leaf because
     // of malformed HTML.
-    RefPtr<Node> node = element();
+    RefPtr<Node> node = this->element();
```
can't this be `= element;` ?

cdumez commented on Source/WebCore/html/RadioInputType.cpp:
```diff
@@ -180,59 +183,62 @@ bool RadioInputType::isKeyboardFocusable(KeyboardEvent* event) const
     // This was done to match the behavior of other browsers.
-    ASSERT(element());
-    return element()->checked();
+    ASSERT(protectedElement());
```
This can remain `element()`

abigailfox replied:
For just the assert or the return as well?

rniwa replied:
Both I guess. Since checked is a trivial function, there is no need to keep the element alive in this case.

abigailfox committed:
Add more smart pointers to RadioInputType.cpp
https://bugs.webkit.org/show_bug.cgi?id=272118
rdar://125877697

Reviewed by Ryosuke Niwa.

Add more smart pointers to RadioInputType.cpp

* Source/WebCore/html/RadioInputType.cpp:
(WebCore::RadioInputType::valueMissing const):
(WebCore::RadioInputType::willUpdateCheckedness):
(WebCore::RadioInputType::handleKeydownEvent):
(WebCore::RadioInputType::handleKeyupEvent):
(WebCore::RadioInputType::isKeyboardFocusable const):
(WebCore::RadioInputType::shouldSendChangeEventAfterCheckedChanged):
(WebCore::RadioInputType::willDispatchClick):
(WebCore::RadioInputType::didDispatchClick):
(WebCore::RadioInputType::matchesIndeterminatePseudoClass const):

Canonical link: https://commits.webkit.org/277319@main
